clang/docs/ReleaseNotes.rst

```diff
@@ -171,6 +171,11 @@ Attribute Changes in Clang
   automatic diagnostic to use parameters of types that the format style
   supports but that are never the result of default argument promotion, such as
   ``float``. (`#59824: <https://github.com/llvm/llvm-project/issues/59824>`_)
+- The ``constructor`` and ``destructor`` attributes now diagnose when:
+  - the priority is not between 101 and 65535, inclusive,
+  - the function it is applied to accepts arguments or has a non-void return
```

Comment:
I don't think the 'non-void' return type should be diagnosed unless it is 'nodiscard'.  Marking something like 'closeAllHandlesAndReturnAStatusIIntendToIgnore` as a destructor isn't harmful IMO.

Reply:
I was on the fence about diagnosing the return value, but ultimately decided it was a type error to have a non-void return value because of calling conventions. e.g., 
```
struct S {
  long double ld1, ld2;
};

__attribute__((constructor))
struct S func(void) {
  return (struct S){ 1.0l, 2.0l };
}
```
where the returned struct is actually returned via a hidden pointer passed to the function. I don't want to play whack-a-mole with signature checking unless there's some compelling use case in the wild where people actually are using these functions to obtain a value outside of ctor/dtor calls.

Reply:
I'm more considering a case where someone wants to use an otherwise 'normal' function as their constructor/destructor, where the return value is perhaps useful if the program is continuing (but then obviously not when it is not).  I would think that the RETURN should at minimum be only a warning (or perhaps warning-as-error?), since this is otherwise a breaking change.

The params list being nothing makes sense, because there is no valid code generation that did that before AFAIK, but an ignorable return type is meaningful.  In your `func` case, I don't see a problem with that, presume the 1st line was something like:
` if (!GlobalAllocator) InitAllocator(GlobalAllocator); `
and they were using the `constructor` call to initialize that (and not caring about the object it was returning).  This would cause this error to be breaking.

Reply:
> I'm more considering a case where someone wants to use an otherwise 'normal' function as their constructor/destructor, where the return value is perhaps useful if the program is continuing (but then obviously not when it is not). I would think that the RETURN should at minimum be only a warning (or perhaps warning-as-error?), since this is otherwise a breaking change.

That was the case that concerned me as well, but I've not seen any evidence of such code existing in the wild.

> The params list being nothing makes sense, because there is no valid code generation that did that before AFAIK, but an ignorable return type is meaningful. In your func case, I don't see a problem with that, presume the 1st line was something like:
> ```
> if (!GlobalAllocator) InitAllocator(GlobalAllocator);
> ```
> and they were using the constructor call to initialize that (and not caring about the object it was returning). This would cause this error to be breaking.

Look at the codegen for my example and that should make it more clear why it's an issue: https://godbolt.org/z/qncWqqovb

The problem is that there's no way to return from that function without inducing UB. You either return a value (and the hidden pointer is invalid so you just did an out of bounds write somewhere) or you don't return any value (no out of bounds write to the hidden pointer, but oops, UB to fall off the end of the function that way).

You can devise circumstances where it's safe, but unless there's evidence that people use the attribute this way, I don't see a compelling reason to give people a gun to point at their feet.

Reply:
Okay, after chatting with @erichkeane about this off-list, it seems we're going to have to support `int` as a return type as there is enough existing code in the wild using it and that returned value is safe on every ABI and calling convention we support. e.g.,

https://searchcode.com/file/664462906/src/sound/alure/CMakeLists.txt (see line 70)
https://searchcode.com/file/624978895/deps/rocksdb/utilities/transactions/lock/range/range_tree/lib/portability/memory.h/ (see line 63)
(and others)

As best we can find, it seems all of these uses are returning `int` specifically. So the current thinking is that we accept `void`, `int`, and `unsigned int` as valid return types, and error on any other type. I can update the documentation for the attributes to explain why.

Reply:
Agreed, I've convinced myself that `int` is really the only currently used example (including the CMake tests, which are particularly troublesome to me), and that we should continue supporting it without diagnostic (again, because CMake).  Everything else is either problematic due to calling-convention issues or at least not used.

So I'm OK with this being a breaking-change error except for `int`.  We've also agreed that `unsigned` makes sense as well because 'why not'.